### ISSUE TITLE
Issues and Questions about Execution of LLaMA using NNTrainer

### ISSUE THREAD

**Issue by Deeksha-20-99** [nnstreamer/nntrainer#2561]
- We have executed the LLaMA model (downloaded from HuggingFace[https://huggingface.co/meta-llama/Llama-2-7b-chat-hf])  using the NNTrainer and obtained the following output by following these steps:

1. File changes made before running the LLaMA model -file path: nntrainer/Applications/LLaMA/PyTorch run the llama_weights_converter.py file to generate the "./llama_fp16.bin" files. (Hugging face LLaMA model) Save the file in the nntrainer/jni directory.
file path: nntrainer/Applications/LLaMA/jni/main.cpp add #define ENABLE_ENCODER2 in the beginning
file path:nntrainer/meson.build add "message ('platform: @0@'.format(get_option('platform')))" in the 28th line of the code. add "message ('enable-fp16: @0@'.format(get_option('enable-fp16')))" in the 68th line of the code
file path:nntrainer/meson_options.txt -enable the fp16 option as true in the 39th line "option('enable-fp16', type: 'boolean', value: true)"
2. Run the "meson build" and "ninja -C build" command in the NNTrainer directory
3. enter the jni directory inside NNTrainer and run "../build/Applications/LLaMA/jni/nntrainer_llama"

- We executed with setting the locale to : “std::locale::global(std::locale("ko_KR.UTF-8”));” got the following output

<img width="728" alt="Korean locale" src="https://github.com/nnstreamer/nntrainer/assets/112724341/5137e97d-fb1f-4c26-996f-50e3b29c888f">

- We executed with setting the locale to : “std::locale::global(std::locale("en_US.UTF-8”));”  got the following output

<img width="676" alt="Screenshot 2024-04-29 at 4 26 29 PM" src="https://github.com/nnstreamer/nntrainer/assets/112724341/6de94d89-1ea9-4f30-9169-431824e9cc6d">

- We executed with commenting the locale statement and got this output

<img width="695" alt="Screenshot 2024-04-29 at 4 30 13 PM" src="https://github.com/nnstreamer/nntrainer/assets/112724341/713a3e25-106d-49f6-8dea-32fc78e84b50">

- Here we are not able to find the correlation between the input and output sequence, hence we wanted to check the way we can infer the results. With setting the locale we are encountering the segmentation error and wanted to know what could be done to resolve this.

- Do you have any recommendation for benchmarks to run to test results from LLaMA execution using NNTrainer?

- We also wanted to ask if we could run NNTrainer on a commercial off-the-shelf GPU. We currently have the NVIDIA A 6000.

Progress update by - Professor Hokeun Kim (https://github.com/hokeun) and his student Deeksha Prahlad (https://github.com/Deeksha-20-99)

**Comment by myungjoo** [comment]
> We also wanted to ask if we could run NNTrainer on a commercial off-the-shelf GPU. We currently have the NVIDIA A 6000.

GPU support of NNTrainer is WIP. I expect to see running LLMs on GPU around May~June. (e.g., https://github.com/nnstreamer/nntrainer/pull/2535 )

This is based on OpenCL because we target GPUs of embedded devices (mobile, TV, home appliances, ...), not servers with powerful A100/H100/B100.

As long as they support OpenCL, they would work; however, not as efficient as CUDA on NVidia GPUs.

**Comment by myungjoo** [comment]
> Do you have any recommendation for benchmarks to run to test results from LLaMA execution using NNTrainer?

Must-have metric: peak memory concumption, first-token latency, per-token latency after the first token output (or "throughput")
Good-to-have metric: energy consumption (J) per given number of input tokens, throughput with given power (W) and thermal budgets, computation resource (CPU, GPU) utilization statistics, average and peak memory (DRAM) traffic.

**Comment by myungjoo** [comment]
> Here we are not able to find the correlation between the input and output sequence, hence we wanted to check the way we can infer the results. With setting the locale we are encountering the segmentation error and wanted to know what could be done to resolve this.

@lhs8928 @baek2sm ?

**Comment by Deeksha-20-99** [comment]
We would like to thank the team for fixing the issue through the commit. We were able to overcome the segmentation fault and run the LLaMA model. We got the output as seen in the images but we are still not able to understand the output that is printed.
<img width="1469" alt="Screenshot 2024-04-30 at 5 36 46 PM" src="https://github.com/nnstreamer/nntrainer/assets/112724341/8681947c-6a9d-409a-915f-f5e1995e84a7">
<img width="1466" alt="Screenshot 2024-04-30 at 5 39 11 PM" src="https://github.com/nnstreamer/nntrainer/assets/112724341/01fb138e-da21-4568-8bc8-8ae8cff76d64">

**Comment by jijoongmoon** [comment]
I wonder whether you changed the configuration for the 7b in HuggingFace. The current implementation is for the 1B.

**Comment by Deeksha-20-99** [comment]
Thank you for the clarification. We have been using the "meta-llama/Llama-2-7b-chat-hf", which is 7B.  We planned to change the model to "TinyLlama/TinyLlama-1.1B-Chat-v1.0", is this the recommended one? If not is there any recommended model to be used for the LLaMA application?

**Comment by jijoongmoon** [comment]
We will check the model including TimyLlam. The current implementation is for the kind of tasks like summarization, tone conversion, etc. and It seems like it does not have tokenizer compatibility with our implementation.  Let us check and we will let you know.

**Comment by baek2sm** [comment]
Currently, the issue of LLaMA application has been modified, and please pull up the latest version and use it.
When you build, you need to set "enable_encoder" option to true in meson_option.txt for using tokenizer.
Then, run with vocab.json and merges.txt files in the same path as the executable file.

**Comment by Deeksha-20-99** [comment]
Thank you for the update. We will look into the modified version.

**Comment by Deeksha-20-99** [comment]
We were able to follow the instructions as mentioned and able to build. We wanted to get the llama model weights to check the model weight file "./llama_fp16.bin", where could we correctly find this file? or which model could we correctly use to get the weights?

**Comment by martinkorelic** [comment]
Hi @myungjoo @jijoongmoon, I'm completely new to this framework, and I want to thank you for your contributions to open source.
My questions are regarding LLMs in this framework:

- If it is supported to fine-tune such a pretrained model already with the given example?
- Will there be more model structures added, or is it possible to start creating custom ones and loading weights from them?
- Is it possible to implement parameter efficient finetuning like LoRA with them?
